File: TLS Inspector/View Controllers/AboutViewController.m
#import "AboutViewController.h"
#import "RecentDomains.h"
#import "GTAppLinks.h"
#import "TitleValueTableViewCell.h"

@interface AboutViewController ()

@property (strong, nonatomic) GTAppLinks * appLinks;

@end

@implementation AboutViewController

static NSString * PROJECT_GITHUB_URL = @"https://github.com/certificate-helper/TLS-Inspector/";
static NSString * PROJECT_URL = @"https://tlsinspector.com/";
static NSString * PROJECT_CONTRIBUTE_URL = @"https://github.com/certificate-helper/TLS-inspector/blob/master/CONTRIBUTE.md";
static NSString * PROJECT_TESTFLIGHT_APPLICATION = @"https://tlsinspector.com/beta.html";

- (void) viewDidLoad {
    [super viewDidLoad];
    self.appLinks = [GTAppLinks new];
}

- (void) didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

# pragma mark - Table View Source

- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView {
    return 3;
}

- (NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    switch (section) {
        case 0:
            return 3;
        case 1:
            return 3;
        case 2:
            return 2;
    }
    return 0;
}

- (UITableViewCell *) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 0) {
        if (indexPath.row == 0) {
            UITableViewCell * switchCell = [tableView dequeueReusableCellWithIdentifier:@"switch" forIndexPath:indexPath];
            UILabel * label = (UILabel *)[switchCell viewWithTag:10];
            label.text = l(@"Remember Recent Lookups");
            label.textColor = themeTextColor;
            UISwitch * toggle = (UISwitch *)[switchCell viewWithTag:20];
            [toggle setOn:[RecentDomains sharedInstance].saveRecentDomains];
            [toggle addTarget:self action:@selector(recentSwitch:) forControlEvents:UIControlEventTouchUpInside];
            return switchCell;
        } else if (indexPath.row == 1) {
            UITableViewCell * switchCell = [tableView dequeueReusableCellWithIdentifier:@"switch" forIndexPath:indexPath];
            UILabel * label = (UILabel *)[switchCell viewWithTag:10];
            label.text = l(@"Show Tips");
            label.textColor = themeTextColor;
            UISwitch * toggle = (UISwitch *)[switchCell viewWithTag:20];
            [toggle setOn:![AppDefaults boolForKey:HIDE_TIPS]];
            [toggle addTarget:self action:@selector(tipsSwitch:) forControlEvents:UIControlEventTouchUpInside];
            return switchCell;
        } else if (indexPath.row == 2) {
            UITableViewCell * toggleCell = [tableView dequeueReusableCellWithIdentifier:@"toggle" forIndexPath:indexPath];
            UILabel * label = (UILabel *)[toggleCell viewWithTag:10];
            label.text = l(@"Theme");
            label.textColor = themeTextColor;
            UISegmentedControl * segment = (UISegmentedControl *)[toggleCell viewWithTag:20];
            [segment setTitle:[lang key:@"Dark"] forSegmentAtIndex:0];
            [segment setTitle:[lang key:@"Light"] forSegmentAtIndex:1];
            if ([AppDefaults boolForKey:USE_LIGHT_THEME]) {
                [segment setSelectedSegmentIndex:1];
            } else {
                [segment setSelectedSegmentIndex:0];
            }
            [segment addTarget:self action:@selector(themeSwitch:) forControlEvents:UIControlEventValueChanged];
            return toggleCell;
        }
    } else if (indexPath.section == 1) {
        UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:@"action" forIndexPath:indexPath];
        UILabel * label = (UILabel *)[cell viewWithTag:1];
        label.textColor = themeTextColor;
        switch (indexPath.row) {
            case 0:
                label.text = l(@"Tell Friends About TLS Inspector");
                break;
            case 1:
                label.text = l(@"Rate TLS Inspector in the App Store");
                break;
            case 2:
                label.text = l(@"Submit Feedback");
                break;
        }
        return cell;
    } else if (indexPath.section == 2) {
        UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:@"action" forIndexPath:indexPath];
        UILabel * label = (UILabel *)[cell viewWithTag:1];
        label.textColor = themeTextColor;
        switch (indexPath.row) {
            case 0:
                label.text = l(@"Contribute on GitHub");
                break;
            case 1:
                label.text = l(@"Test New Features");
                break;
        }
        return cell;
    }
    return nil;
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 1 && indexPath.row == 0) {
        NSString * blurb = format(@"Trust & Safety On-the-go with TLS Inspector: %@", PROJECT_URL);
        UIActivityViewController *activityController = [[UIActivityViewController alloc]
                                                        initWithActivityItems:@[blurb]
                                                        applicationActivities:nil];
        activityController.popoverPresentationController.sourceView = [tableView cellForRowAtIndexPath:indexPath];
        [self presentViewController:activityController animated:YES completion:nil];
    } else if (indexPath.section == 1 && indexPath.row == 1) {
        [self.appLinks showAppInAppStore:1100539810 inViewController:self dismissed:^{
            //
        }];
    } else if (indexPath.section == 1 && indexPath.row == 2) {
        [uihelper
         presentActionSheetInViewController:self
         attachToTarget:[ActionTipTarget targetWithView:[tableView cellForRowAtIndexPath:indexPath]]
         title:l(@"What kind of feedback would you like to submit?")
         subtitle:l(@"All feedback is appreciated!")
         cancelButtonTitle:l(@"Cancel")
         items:@[
                 l(@"Report Issue on GitHub"),
                 l(@"Contact Us")
                 ]
         dismissed:^(NSInteger selectedIndex) {
             switch (selectedIndex) {
                 case 0:
                     open_url(nstrcat(PROJECT_GITHUB_URL, @"issues/new"));
                     break;
                 case 1: {
                     [self.appLinks
                      showEmailComposeSheetForApp:@"TLS Inspector"
                      email:@"'TLS Inspector Project Manager' <tls-inspector@ecnepsnai.com>"
                      inViewController:self dismissed:^{
                         //
                     }];
                     break;
                 } default:
                     break;
             }
         }];
    } else if (indexPath.section == 2 && indexPath.row == 0) {
        open_url(PROJECT_CONTRIBUTE_URL);
    } else if (indexPath.section == 2 && indexPath.row == 1) {
        open_url(PROJECT_TESTFLIGHT_APPLICATION);
    }
}

- (NSString *) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    switch (section) {
        case 0:
            return l(@"Options");
        case 1:
            return l(@"Share & Feedback");
        case 2:
            return l(@"Get Involved");
    }
    return @"";
}

- (NSString *) tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section {
    switch (section) {
        case 1: {
            NSDictionary * infoDictionary = [[NSBundle mainBundle] infoDictionary];
            return format(@"%@ (%@). %@, %@",
                          [infoDictionary objectForKey:@"CFBundleShortVersionString"],
                          [infoDictionary objectForKey:(NSString *)kCFBundleVersionKey],
                          [CKCertificate openSSLVersion],
                          [CKServerInfo libcurlVersion]);
        } case 2:
            return l(@"TLS Inspector is Free and Libre software licensed under GNU GPLv3. TLS Inspector is copyright © 2016 Ian Spence.");
    }
    return @"";
}

- (void) recentSwitch:(UISwitch *)sender {
    [RecentDomains sharedInstance].saveRecentDomains = sender.isOn;
}

- (void) tipsSwitch:(UISwitch *)sender {
    [AppDefaults setBool:!sender.isOn forKey:HIDE_TIPS];
}

- (void) themeSwitch:(UISegmentedControl *)sender {
    if (sender.selectedSegmentIndex == 0) {
        [AppDefaults setBool:NO forKey:USE_LIGHT_THEME];
    } else {
        [AppDefaults setBool:YES forKey:USE_LIGHT_THEME];
    }
}

@end

File: TLS Inspector/View Controllers/AboutViewController.swift
import UIKit
import CertificateKit

class AboutTableViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    let projectGithubURL = "https://tlsinspector.com/github.html"
    let projectURL = "https://tlsinspector.com/"
    let projectContributeURL = "https://tlsinspector.com/contribute.html"
    let testflightURL = "https://tlsinspector.com/beta.html"
    let twitterURL = "https://twitter.com/tlsinspector"
    @IBOutlet weak var lockCircle: UIImageView!
    @IBOutlet weak var tableView: UITableView!
    var quotes: [String] = []
    var taps = 0
    var sections: [TableViewSection] = []

    override func viewDidLoad() {
        super.viewDidLoad()

        if let q = loadQuotes() {
            self.quotes = q
        }

        let tap = UITapGestureRecognizer(target: self, action: #selector(self.didTapPicture(target:)))
        self.lockCircle.isUserInteractionEnabled = true
        self.lockCircle.addGestureRecognizer(tap)

        self.sections = [
            buildShareSection(),
            buildGetInvolvedSection()
        ]
    }

    func buildShareSection() -> TableViewSection {
        let section = TableViewSection()
        section.title = lang(key: "Share & Feedback")
        let opensslVersion = CertificateKit.opensslVersion()
        let libcurlVersion = CertificateKit.libcurlVersion()
        section.footer = "App: \(AppInfo.version()) (\(AppInfo.build())), OpenSSL: \(opensslVersion), tiny-curl: \(libcurlVersion)"

        return section
    }

    func buildGetInvolvedSection() -> TableViewSection {
        let section = TableViewSection()
        section.title = lang(key: "Get Involved")
        section.footer = lang(key: "copyright_license_footer")

        return section
    }

    func loadQuotes() -> [String]? {
        guard let quotesPath = Bundle.main.path(forResource: "QuoteList", ofType: "plist") else { return nil }
        guard let quotes = NSArray.init(contentsOfFile: quotesPath) as? [String] else { return nil }
        return quotes
    }

    @objc func didTapPicture(target: UIImageView) {
        if self.quotes.count == 0 {
            return
        }

        self.taps += 1
        if self.taps >= 3 {
            self.taps = 0

            UIHelper(self).presentAlert(title: self.quotes.randomElement() ?? "", body: "", dismissed: nil)
        }
    }

    // MARK: - Table view data source
    func numberOfSections(in tableView: UITableView) -> Int {
        return 2
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 0 {
            return 3
        } else if section == 1 {
            return 3
        }

        return 0
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Basic", for: indexPath)
        if indexPath.section == 0 && indexPath.row == 0 {
            cell.textLabel?.text = lang(key: "Share TLS Inspector")
        } else if indexPath.section == 0 && indexPath.row == 1 {
            cell.textLabel?.text = lang(key: "Rate in App Store")
        } else if indexPath.section == 0 && indexPath.row == 2 {
            cell.textLabel?.text = lang(key: "Provide Feedback")
        } else if indexPath.section == 1 && indexPath.row == 0 {
            cell.textLabel?.text = lang(key: "Contribute to TLS Inspector")
        } else if indexPath.section == 1 && indexPath.row == 1 {
            cell.textLabel?.text = lang(key: "Test New Features")
        } else if indexPath.section == 1 && indexPath.row == 2 {
            return tableView.dequeueReusableCell(withIdentifier: "Twitter", for: indexPath)
        }
        return cell
    }

    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if section == 0 {
            return lang(key: "Share & Feedback")
        } else if section == 1 {
            return lang(key: "Get Involved")
        }

        return nil
    }

    func tableView(_ tableView: UITableView, titleForFooterInSection section: Int) -> String? {
        if section == 0 {
            let opensslVersion = CertificateKit.opensslVersion()
            let libcurlVersion = CertificateKit.libcurlVersion()
            return "App: \(AppInfo.version()) (\(AppInfo.build())), OpenSSL: \(opensslVersion), tiny-curl: \(libcurlVersion)"
        } else if section == 1 {
            return lang(key: "copyright_license_footer")
        }

        return nil
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let cell = tableView.cellForRow(at: indexPath) else { return }

        if indexPath.section == 0 && indexPath.row == 0 {
            let blub = lang(key: "Trust & Safety On-The-Go with TLS Inspector: {url}", args: [projectURL])
            let activityController = UIActivityViewController(activityItems: [blub], applicationActivities: nil)
            ActionTipTarget(view: cell).attach(to: activityController.popoverPresentationController)
            self.present(activityController, animated: true, completion: nil)
        } else if indexPath.section == 0 && indexPath.row == 1 {
            AppLinks.current.showAppStore(self, dismissed: nil)
        } else if indexPath.section == 0 && indexPath.row == 2 {
            ContactTableViewController.show(self) { (support) in
                AppLinks.current.showEmailCompose(viewController: self, object: support, includeLogs: false, dismissed: nil)
            }
        } else if indexPath.section == 1 && indexPath.row == 0 {
            OpenURLInSafari(projectContributeURL)
        } else if indexPath.section == 1 && indexPath.row == 1 {
            OpenURLInSafari(testflightURL)
        } else if indexPath.section == 1 && indexPath.row == 2 {
            OpenURLInSafari(twitterURL)
        }
    }
}
